File: items/wallet.js
const prompt = require("prompt-sync")(); 


class IDCard {
    constructor(name, number, lost) {
        this.name = name;
        this.number = number;
        this.lost = lost;
    }
}

class Money {
    constructor(currency, amount) {
        this.currency = currency;
        this.amount = amount;
    }
}

class CreditCard {
    constructor(owner, number, expirationDate) {
        this.owner = owner;
        this.number = number;
        this.expirationDate = expirationDate;
    }
}

class DrivingLicence {
    constructor(owner, category) {
        this.owner = owner;
        this.category = category;
    }
}

class IDPhoto {
    constructor(owner) {
        this.owner = owner;
    }
}

class BusinessCard {
    constructor(owner, companyName) {
        this.owner = owner;
        this.companyName = companyName;
    }
}

class Wallet {
    constructor(owner, isLost, idCards, moneyList, creditCards, drivingLicences, idPhotos, businessCards) {
        this.owner = owner;
        this.isLostStatus = isLost;
        this.idCards = idCards;
        this.moneyList = moneyList;
        this.creditCards = creditCards;
        this.drivingLicences = drivingLicences;
        this.idPhotos = idPhotos;
        this.businessCards = businessCards;
    }

    getOwner() {
        return this.owner;
    }

    setLost() {
        this.isLostStatus = !this.isLostStatus;
    }

    getIdCards() {
        return this.idCards;
    }

    addIdCard() {
        if (!this.isLostStatus) {
            const idCardName = prompt("Enter ID Card name: ");

            const idCardNumber = prompt("Enter ID Card number: ");

            const newIdCard = new IDCard(idCardName, idCardNumber, false);

            this.idCards.push(newIdCard);

            console.log("ID Card added to the wallet.");
        } else {
            console.log("Cannot add ID Card. The wallet is lost.");
        }
    }

    getMoneyList() {
        return this.moneyList;
    }

    addMoney() {
        if (!this.isLostStatus) {
            const amount = parseFloat(prompt("Enter the amount of money: "));

            console.log("Choose currency:");
            console.log("\n1. Dollars\n2. Euro\n3. Ariary");
            console.log();
            const currencyChoice = parseInt(prompt("\nEnter currency choice (1-3): "));

            let currency;
            switch (currencyChoice) {
                case 1:
                    currency = "Dollars";
                    break;
                case 2:
                    currency = "Euro";
                    break;
                case 3:
                    currency = "Ariary";
                    break;
                default:
                    console.log("Invalid currency choice. Using Dollars by default.");
                    currency = "Dollars";
            }

            const money = new Money(currency, amount);
            this.moneyList.push(money);
            console.log("Money added to the wallet.");
        } else {
            console.log("Cannot add money. The wallet is lost.");
        }
    }

    getCreditCards() {
        return this.creditCards;
    }

    addCreditCard() {
        if (!this.isLostStatus) {
            const cardOwner = prompt("Enter Credit Card owner: ");

            const cardNumber = prompt("Enter Credit Card number: ");

            const dateString = prompt("Enter expiration date (yyyy-MM-dd): ");
            const expirationDate = new Date(dateString);

            const newCreditCard = new CreditCard(cardOwner, cardNumber, expirationDate);
            this.creditCards.push(newCreditCard);

            console.log("Credit Card added to the wallet.");
        } else {
            console.log("Cannot add Credit Card. The wallet is lost.");
        }
    }

    getDrivingLicences() {
        return this.drivingLicences;
    }

    addDrivingLicence() {
        if (!this.isLostStatus) {
            const licenceOwner = prompt("Enter Driving Licence owner: ");

            console.log("Choose Driving Licence category:");
            console.log("1. A");
            console.log("2. B");
            console.log("3. C");
            console.log("4. D");
            console.log("5. E");
            const categoryChoice = parseInt(prompt("Enter category choice (1-5): "));

            let licenceCategory;
            switch (categoryChoice) {
                case 1:
                    licenceCategory = "A";
                    break;
                case 2:
                    licenceCategory = "B";
                    break;
                case 3:
                    licenceCategory = "C";
                    break;
                case 4:
                    licenceCategory = "D";
                    break;
                case 5:
                    licenceCategory = "E";
                    break;
                default:
                    console.log("Invalid category choice. Using A by default.");
                    licenceCategory = "A";
            }

            const newDrivingLicence = new DrivingLicence(licenceOwner, licenceCategory);
            this.drivingLicences.push(newDrivingLicence);

            console.log("Driving Licence added to the wallet.");
        } else {
            console.log("Cannot add Driving Licence. The wallet is lost.");
        }
    }

    getIdPhotos() {
        return this.idPhotos;
    }

    addIdPhoto() {
        if (!this.isLostStatus) {
            const photoOwner = prompt("Enter ID Photo owner: ");

            const newIdPhoto = new IDPhoto(photoOwner);
            this.idPhotos.push(newIdPhoto);

            console.log("ID Photo added to the wallet.");
        } else {
            console.log("Cannot add ID Photo. The wallet is lost.");
        }
    }

    getBusinessCards() {
        return this.businessCards;
    }

    addBusinessCard() {
        if (!this.isLostStatus) {
            const cardOwner = prompt("Enter Business Card owner: ");

            const companyName = prompt("Enter Company Name: ");

            const newBusinessCard = new BusinessCard(cardOwner, companyName);
            this.businessCards.push(newBusinessCard);

            console.log("Business Card added to the wallet.");
        } else {
            console.log("Cannot add Business Card. The wallet is lost.");
        }
    }

    removeItemByType(itemType) {
        switch (itemType) {
            case "ID_CARD":
                this.removeItem(this.getIdCards(), "ID Card");
                break;
            case "MONEY":
                this.removeItem(this.getMoneyList(), "Money item");
                break;
            case "CREDIT_CARD":
                this.removeItem(this.getCreditCards(), "Credit Card");
                break;
            case "BUSINESS_CARD":
                this.removeItem(this.getBusinessCards(), "Business Card");
                break;
            case "DRIVING_LICENCE":
                this.removeItem(this.getDrivingLicences(), "Driving Licence");
                break;
            case "ID_PHOTO":
                this.removeItem(this.getIdPhotos(), "ID Photo");
                break;
        }
    }
    
    removeItem(itemList, itemType) {
        if (itemList.length === 0) {
            console.log("\nNo " + itemType + "s found in the wallet.");
            return;
        }
    
        console.log("\n" + itemType + "s in the wallet:");
        for (let i = 0; i < itemList.length; i++) {
            console.log(`\n${i + 1}. ${JSON.stringify(itemList[i])}`);
        }
    
        const choice = parseInt(prompt("\nEnter the number of the " + itemType + " to remove:"));
    
        if (choice >= 1 && choice <= itemList.length) {
            const removedItem = itemList.splice(choice - 1, 1)[0];
            console.log("\nRemoved " + itemType + ": " + JSON.stringify(removedItem));
        } else {
            console.log("\nInvalid choice. No " + itemType + " removed.");
        }
    }
}

module.exports = Wallet;

File: items/wallet.py
import datetime, json

from items.id_card import IDCard
from items.credit_card import CreditCard
from items.driving_licence import DrivingLicence
from items.id_photo import IDPhoto
from items.business_card import BusinessCard
from items.money import Money

class Wallet:
    def __init__(self, owner, is_lost, id_cards, money_list, credits_cards, driving_licences, id_photos,business_cards):
        self.owner = owner
        self.is_lost_status = is_lost
        self.id_cards = id_cards
        self.money_list = money_list
        self.credit_cards = credits_cards
        self.driving_licences = driving_licences
        self.id_photos = id_photos
        self.business_cards = business_cards

    def get_owner(self):
        return self.owner
    
    def set_lost(self):
        self.is_lost_status = not self.is_lost_status

    def get_id_cards(self):
        return self.id_cards
    
    def add_id_card(self):
        if not self.is_lost_status:
            id_card_name = input("Enter ID Card name: ")
            id_card_number = input("Enter ID Card number: ")
            new_id_card = IDCard(id_card_name, id_card_number, False)
            self.id_cards.append(new_id_card)
            print("ID Card added to the wallet.")
        else:
            print("Cannot add ID Card. The wallet is lost.")

    def get_money_list(self):
        return self.money_list

    def add_money(self):
        if not self.is_lost_status:
            amount = float(input("Enter the amount of money: "))
            print("Choose currency:")
            print("\n1. Dollars\n2. Euro\n3. Ariary")
            currency_choice = int(input("\nEnter currency choice (1-3): "))
            currencies = ["Dollars", "Euro", "Ariary"]
            currency = currencies[currency_choice - 1] if 1 <= currency_choice <= 3 else "Dollars"
            money = Money(currency, amount)
            self.money_list.append(money)
            print("Money added to the wallet.")
        else:
            print("Cannot add money. The wallet is lost.")

    def get_credit_cards(self):
        return self.credit_cards

    def add_credit_card(self):
        if not self.is_lost_status:
            card_owner = input("Enter Credit Card owner: ")
            card_number = input("Enter Credit Card number: ")
            expiration_date_string = input("Enter expiration date (yyyy-MM-dd): ")
            new_credit_card = CreditCard(card_owner, card_number, expiration_date_string)
            self.credit_cards.append(new_credit_card)
            print("Credit Card added to the wallet.")
        else:
            print("Cannot add Credit Card. The wallet is lost.")

    def get_driving_licences(self):
        return self.driving_licences

    def add_driving_licence(self):
        if not self.is_lost_status:
            licence_owner = input("Enter Driving Licence owner: ")
            print("Choose Driving Licence category:")
            print("1. A\n2. B\n3. C\n4. D\n5. E")
            category_choice = int(input("Enter category choice (1-5): "))
            categories = ["A", "B", "C", "D", "E"]
            licence_category = categories[category_choice - 1] if 1 <= category_choice <= 5 else "A"
            new_driving_licence = DrivingLicence(licence_owner, licence_category)
            self.driving_licences.append(new_driving_licence)
            print("Driving Licence added to the wallet.")
        else:
            print("Cannot add Driving Licence. The wallet is lost.")

    def get_id_photos(self):
        return self.id_photos

    def add_id_photo(self):
        if not self.is_lost_status:
            photo_owner = input("Enter ID Photo owner: ")
            new_id_photo = IDPhoto(photo_owner)
            self.id_photos.append(new_id_photo)
            print("ID Photo added to the wallet.")
        else:
            print("Cannot add ID Photo. The wallet is lost.")

    def get_business_cards(self):
        return self.business_cards

    def add_business_card(self):
        if not self.is_lost_status:
            card_owner = input("Enter Business Card owner: ")
            company_name = input("Enter Company Name: ")
            new_business_card = BusinessCard(card_owner, company_name)
            self.business_cards.append(new_business_card)
            print("Business Card added to the wallet.")
        else:
            print("Cannot add Business Card. The wallet is lost.")

    def remove_item_by_type(self, item_type):
        switcher = {
            "ID_CARD": lambda: self.remove_item(self.get_id_cards(), "ID Card"),
            "MONEY": lambda: self.remove_item(self.get_money_list(), "Money item"),
            "CREDIT_CARD": lambda: self.remove_item(self.get_credit_cards(), "Credit Card"),
            "BUSINESS_CARD": lambda: self.remove_item(self.get_business_cards(), "Business Card"),
            "DRIVING_LICENCE": lambda: self.remove_item(self.get_driving_licences(), "Driving Licence"),
            "ID_PHOTO": lambda: self.remove_item(self.get_id_photos(), "ID Photo")
        }
        switcher.get(item_type, lambda: print("Invalid item type."))()

    def remove_item(self, item_list, item_type):
        if not item_list:
            print(f"\nNo {item_type}s found in the wallet.")
            return

        print(f"\n{item_type}s in the wallet:")
        for i, item in enumerate(item_list, start=1):
            print(f"\n{i}. {json.dumps(item.__dict__)}")

        choice = int(input(f"\nEnter the number of the {item_type} to remove: "))

        if 1 <= choice <= len(item_list):
            removed_item = item_list.pop(choice - 1)
            print(f"\nRemoved {item_type}: {json.dumps(removed_item.__dict__)}")
        else:
            print(f"\nInvalid choice. No {item_type} removed.")